<h3>1.- Importación de librerías</h3>

In [133]:
# Importamos las librerías necesarias

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

<h3>2.- Importación del Dataset</h3>

In [134]:
# Importamos el Dataset leyendo el archivo csv en un DataFrame

df_bmw0 = pd.read_csv("dataset/bmw_pricing_v3.csv")

<h3>3.- Exploración Inicial del Dataset</h3>

Ejemplo de columnas

In [135]:
df_bmw0.sample(10)

marca            modelo        km  potencia fecha_registro tipo_gasolina  \
2201   BMW               520  146402.0     137.0     2012-08-01        diesel   
3930   BMW                X3   78087.0     105.0     2013-01-01        diesel   
4572   NaN                X4   54433.0     140.0     2015-06-01        diesel   
3755   NaN                X3   73086.0     135.0            NaN        diesel   
64     BMW               420   61300.0     135.0            NaN        diesel   
1639   BMW               525  193148.0     155.0     2012-03-01        diesel   
271    NaN               320  236259.0     135.0     2012-01-01        diesel   
1284   BMW               320  101974.0     135.0     2014-06-01        diesel   
1966   BMW  318 Gran Turismo  178834.0     100.0     2014-01-01        diesel   
782    BMW               520  167112.0     135.0     2013-11-01        diesel   

      color tipo_coche volante_regulable aire_acondicionado camara_trasera  \
2201   grey        NaN              True               True          False   
3930  black        NaN             False               True          False   
4572  white        NaN             False              False          False   
3755  black        NaN             False              False          False   
64      NaN        NaN              True               True          False   
1639   grey     estate              True               True          False   
271    blue     estate             False              False          False   
1284   grey     estate             False              False          False   
1966    NaN        NaN              True               True          False   
782   black        NaN              True               True          False   

     asientos_traseros_plegables elevalunas_electrico bluetooth    gps  \
2201                         NaN                False     False   True   
3930                         NaN                False     False   True   
4572                         NaN                False     False  False   
3755                       False                False     False  False   
64                           NaN                 True     False   True   
1639                         NaN                 True     False   True   
271                          NaN                False     False   True   
1284                         NaN                False     False   True   
1966                         NaN                 True     False   True   
782                          NaN                 True     False   True   

     alerta_lim_velocidad   precio fecha_venta  
2201                 True  16200.0  2018-05-01  
3930                  NaN  21700.0  2018-02-01  
4572                False  33200.0  2018-07-01  
3755                 True  17500.0  2018-09-01  
64                   True  27700.0  2018-02-01  
1639                False  14100.0  2018-08-01  
271                 False   6400.0  2018-01-01  
1284                  NaN  14900.0  2018-06-01  
1966                False  13100.0  2018-08-01  
782                  True  18900.0  2018-03-01

Número de filas y columnas

In [136]:
df_bmw0.shape

(4843, 18)

Columnas y tipo de Datos

In [137]:
df_bmw0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   marca                        3873 non-null   object 
 1   modelo                       4840 non-null   object 
 2   km                           4841 non-null   float64
 3   potencia                     4842 non-null   float64
 4   fecha_registro               2420 non-null   object 
 5   tipo_gasolina                4838 non-null   object 
 6   color                        4398 non-null   object 
 7   tipo_coche                   3383 non-null   object 
 8   volante_regulable            4839 non-null   object 
 9   aire_acondicionado           4357 non-null   object 
 10  camara_trasera               4841 non-null   object 
 11  asientos_traseros_plegables  1452 non-null   object 
 12  elevalunas_electrico         4841 non-null   object 
 13  bluetooth         

Número de nulos

In [138]:
df_bmw0.isnull().sum()

marca                           970
modelo                            3
km                                2
potencia                          1
fecha_registro                 2423
tipo_gasolina                     5
color                           445
tipo_coche                     1460
volante_regulable                 4
aire_acondicionado              486
camara_trasera                    2
asientos_traseros_plegables    3391
elevalunas_electrico              2
bluetooth                       728
gps                               0
alerta_lim_velocidad            728
precio                            6
fecha_venta                       1
dtype: int64

### Comentarios

 - El Dataset tiene 4843 registros

 - Todas las columnas (excepto la de gps) tienen algún nulo, aunque hay 9 de ellas que tienen solo algunos valores nulos sueltos que podrán imputarse fácilmente.

 - Hay 3 columnas (color, bluetooth y alerta_lim_velocidad) que tienen un número moderado de nulos (entre el 10% y el 20% del total de registros)

 - La columna tipo_coche tiene un 33% de valores nulos

 - La columna fecha_registro tiene un 50% de valores nulos

 - La columna asientos_traseros_plegables tiene un 70% de valores nulos

 - Tenemos 3 columnas numéricas (km, potencia y precio), 1 columna booleana (gps), y el resto de las columnas son objects (strings)

 - A la vista de los datos, fecha_registro y fecha_venta podrían transformarse a tipo Date

 - A la vista de los datos, las columnas volante_regulable, aire-acondicionado, camara_trasera, asientos_traseros_plegables, 
elevalunas_electrico, bluetooh, alerta_lim_velocidad podrían transformarse a tipo Booleano.

<h3>4.- Comprobación de registros duplicados</h3>

In [139]:
df_bmw0[df_bmw0.duplicated()].shape

(0, 18)

No hay 2 registros completamente idénticos, aunque eso no descartaría por completo los duplicados (podría haber registros iguales donde se haya modificado solamente el valor de alguna de las columnas). Como comprobación rápida, vamos a ver si hay coches de la misma marca y modelo, con el mismo kilometraje, y que tengan la misma fecha de registro y de venta (si los hay, podría ser que se hubiera introdcido el mismo registro por duplicado, aunque no todos los datos sean idénticos).

In [140]:
df_bmw0[df_bmw0[['marca','modelo','km','fecha_registro', 'fecha_venta']].duplicated(keep=False)].shape

(0, 18)

A la vista de los resultados, podemos deducir que no hay registros duplicados.

<h3>5.- Análisis de nulos</h3>

In [141]:
# Como vamos a empezar a realizar cambios, hacemos una copia del DataFrame y empezamos a hacer un análisis de los nulos

df_bmw1 = df_bmw0.copy()
df_bmw1.isnull().sum()

marca                           970
modelo                            3
km                                2
potencia                          1
fecha_registro                 2423
tipo_gasolina                     5
color                           445
tipo_coche                     1460
volante_regulable                 4
aire_acondicionado              486
camara_trasera                    2
asientos_traseros_plegables    3391
elevalunas_electrico              2
bluetooth                       728
gps                               0
alerta_lim_velocidad            728
precio                            6
fecha_venta                       1
dtype: int64

Analizamos cada columna independientemente

### 5.1 Columna Marca

In [142]:
# La columna 'marca' es de tipo object (string). Hacemos un value_counts() de esa columna para ver la distribución de valores, y cuáto representan los nulos

df_bmw1['marca'].value_counts(dropna=False, normalize=True)*100

marca
BMW    79.971092
NaN    20.028908
Name: proportion, dtype: float64

El 80% de los registros son de la marca BMW, y tenemos un 20% de nulos. A priori, esta columna no agrega información (ya que se supone que todos los coches son BMW). Vamos a hacer una comprobación, para ver si podemso imputar la marca BMW a algunos de los modelos donde la marca es null. Para ello, vamos a obtener una lista de modelos don el campo marca es BMW, y otra lista donde el campo marca es null, y las compararemos. Los modelos que se encuentren en ambas listas, podemos suponer que son de BMW.

In [174]:
#Creamos una lista con los modelos que sí tienen Marca (BMW)

lista_modelos_con_marca = df_bmw1[df_bmw1['marca'].notnull()]['modelo'].unique().tolist()
len(lista_modelos_con_marca)

73

In [175]:
#Creamos otra lista, con los modelos que no tienen marca 

lista_modelos_sin_marca = df_bmw1[df_bmw1['marca'].isnull()]['modelo'].unique().tolist()
len(lista_modelos_sin_marca)

51

In [176]:
#Creamos una lista con los índices de los elementos de la lista de modelos sin marca que queremos eliminar (los que ya se encuentran en la lista de modelos con marca BMW)

indices_a_eliminar = []

for modelo in lista_modelos_sin_marca:
    if modelo in lista_modelos_con_marca:
        indices_a_eliminar.append(lista_modelos_sin_marca.index(modelo))

In [177]:
# Si intentamos eliminar esos índices de la lista de modelos sin marca, tendremos un error, ya que al eliminar un elemento, todos los demas se desplazan una posición a la izquierda modificando su índice.
# Para evitarlo, le daremos la vuelta a la lista e iremos elimianndo de atrás hacia adelante.

indices_a_eliminar.sort(reverse=True)

# Eliminar los elementos de la lista de modelos sin marca, basados en los índices proporcionados
for indice in indices_a_eliminar:
    if indice < len(lista_modelos_sin_marca):
        del lista_modelos_sin_marca[indice]

In [184]:
# Comprobamos los modelos que no tienen asociada la marca BMV
lista_modelos_sin_marca

['M235', '630', '220 Active Tourer', '225 Active Tourer']

In [185]:
# Vemos cuántos registros son en el DataFrame

df_bmw1[df_bmw1.modelo.isin(lista_modelos_sin_marca)]

marca             modelo        km  potencia fecha_registro  \
57     NaN               M235   34082.0     240.0            NaN   
84     NaN               M235   69736.0     240.0            NaN   
102    NaN               M235   37619.0     240.0            NaN   
107    NaN                630  149943.0     190.0            NaN   
4804   NaN  220 Active Tourer  153358.0     140.0     2015-06-01   
4825   NaN  225 Active Tourer   33300.0     170.0            NaN   

     tipo_gasolina  color tipo_coche volante_regulable aire_acondicionado  \
57          petrol   grey      coupe             False               True   
84          petrol   grey        NaN              True               True   
102         petrol  black        NaN              True              False   
107         petrol  black      coupe              True               True   
4804        diesel   grey        van             False                NaN   
4825        petrol    NaN        van              True               True   

     camara_trasera asientos_traseros_plegables elevalunas_electrico  \
57             True                       False                 True   
84             True                       False                 True   
102           False                         NaN                 True   
107            True                       False                 True   
4804          False                         NaN                False   
4825          False                         NaN                False   

     bluetooth    gps alerta_lim_velocidad   precio fecha_venta  
57        True   True                  NaN  30200.0  2018-01-01  
84        True   True                False  30000.0  2018-05-01  
102       True   True                 True  27700.0  2018-05-01  
107       True   True                 True  10500.0  2018-05-01  
4804     False   True                 True  11700.0  2018-02-01  
4825     False  False                 True  19200.0  2018-05-01

### Conclusiones sobre la marca

De todos los registros que no tienen marca, los que tienen modelo la mayoría corresponde con un modelo de BMW, por lo que podría imputarse que la marca es BMW.
Solamente hay 6 registros sin marca, cuyo modelo no se encuentra entre los registros que sí tienen marca. 
En cualquier caso, dado que esa columna tiene casi exclusivamente un valor, no aporta información, por lo que la eliminaremos del Dataset.
